### PR TITLE
funnel charts

### PR DESCRIPTION
**Pull request by dariuszlacheta**
#114

### CODE REVIEW

File: lib/libraries/default/funnel-3d.js [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+            prevElemWidth = prevElemWidth * d[1]/matrix[i][1];
+            percent = ((d[1]/matrix[i][1])*100).toFixed(percentsDecimals);
+          }
+          if(percentsCounting === 'absolute'){
+            prevElemWidth = svgWidth * d[1]/matrix[1][1];
+            percent = ((d[1]/matrix[1][1])*100).toFixed(percentsDecimals);
+          }
+        }
+        let label = d[0];
+        if(Object.keys(labelMapping).length > 0){
```
Comment (apryka) [inline lib/libraries/default/funnel-3d.js]: please update with `if(Object.keys(labelMapping).length)`

File: lib/libraries/default/funnel-3d.js [hunk 2]
```diff
@@ -0,0 +1,347 @@
...
+          for (var key in labelMapping) {
```
Comment (apryka) [inline lib/libraries/default/funnel-3d.js]: `var` -> `let`

File: lib/libraries/default/funnel-3d.js [hunk 3]
```diff
@@ -0,0 +1,347 @@
...
+import { textWrap } from '../../utils/svg-text-wrap';
+import { prettyNumber } from '../../utils/pretty-number';
+
+export default class Funnel3d {
+  render() {
+    const { matrix } = this.dataset;
+    const {
+      colors,
+      container,
+      theme,
```
Comment (apryka) [inline lib/libraries/default/funnel-3d.js]: do we need this `theme` const? it is not used anywhere

File: lib/libraries/default/funnel-3d.js [hunk 4]
```diff
@@ -0,0 +1,347 @@
...
+          - margin.top - margin.bottom - 30;
+    const elemHeight = svgHeight/(matrix.length - 1);
+    const yMarginElement = elemHeight * 0.3;
+    const shadowCut = (yMarginElement * 2.5);
+    let prevElemWidth = svgWidth;
+    let percent = (100).toFixed(percentsDecimals);
+
+    //normal funnel polygons preparing
+    const polygons = matrix.slice(1).map((d, i) => {
+        let newPoints = [
```
Comment (apryka) [inline lib/libraries/default/funnel-3d.js]: should be `const`. newPoints is not re-assigned

File: lib/libraries/default/funnel-3d.js [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+          }
+        }
+        let result = d[1];
+        if ( (typeof opts['prettyNumber'] === 'undefined' || opts['prettyNumber'] === true)
+          && !isNaN(parseInt(d[1])) ) {
+            result = prettyNumber(d[1]);
+        }
+        return {
+          name: d[0],
+          label: label,
```
Comment (apryka) [inline lib/libraries/default/funnel-3d.js]: just `label` instead of `label:label`

File: lib/libraries/default/funnel-3d.js [hunk 6]
```diff
@@ -0,0 +1,347 @@
...
+          percent: percent + '%',
+          result: result,
```
Comment (apryka) [inline lib/libraries/default/funnel-3d.js]: just `result`

File: lib/libraries/default/funnel-3d.js [hunk 7]
```diff
@@ -0,0 +1,347 @@
...
+      container,
+      theme,
+      labelMapping,
+      colorMapping
+    } = this.config;
+    const opts = this.config;
+    const {
+      percents,
+      lines,
+      spacesBetweenElements,
```
Comment (apryka) [inline lib/libraries/default/funnel-3d.js]: not used here

File: lib/libraries/default/horizontal-funnel-3d.js [hunk 8]
```diff
@@ -0,0 +1,341 @@
...
+      container,
+      theme,
+      labelMapping,
+      colorMapping
+    } = this.config;
+    const opts = this.config;
+    const {
+      percents,
+      lines,
+      spacesBetweenElements,
```
Comment (apryka) [inline lib/libraries/default/horizontal-funnel-3d.js]: not used here

File: lib/libraries/default/funnel.js [hunk 9]
```diff
@@ -0,0 +1,305 @@
...
+import { textWrap } from '../../utils/svg-text-wrap';
+import { prettyNumber } from '../../utils/pretty-number';
+
+export default class Funnel {
+  render() {
+    const { matrix } = this.dataset;
+    const {
+      colors,
+      container,
+      theme,
```
Comment (apryka) [inline lib/libraries/default/funnel.js]: not used anywhere

File: lib/libraries/default/horizontal-funnel-3d.js [hunk 10]
```diff
@@ -0,0 +1,341 @@
...
+import { textWrap } from '../../utils/svg-text-wrap';
+import { prettyNumber } from '../../utils/pretty-number';
+
+export default class HorizontalFunnel3d {
+  render() {
+    const { matrix } = this.dataset;
+    const {
+      colors,
+      container,
+      theme,
```
Comment (apryka) [inline lib/libraries/default/horizontal-funnel-3d.js]: not used anywhere

File: lib/libraries/default/horizontal-funnel.js [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+import { textWrap } from '../../utils/svg-text-wrap';
+import { prettyNumber } from '../../utils/pretty-number';
+
+export default class HorizontalFunnel {
+  render() {
+    const { matrix } = this.dataset;
+    const {
+      colors,
+      container,
+      theme,
```
Comment (apryka) [inline lib/libraries/default/horizontal-funnel.js]: not used anywhere

File: lib/utils/svg-text-wrap.js [hunk 12]
```diff
@@ -0,0 +1,34 @@
+import * as d3 from 'd3';
+
+export function textWrap(text, width) {
+  text.each(function () {
+    var text = d3.select(this),
```
Comment (apryka) [inline lib/utils/svg-text-wrap.js]: `var` to `let`

File: lib/libraries/default/funnel-3d.js [hunk 13]
```diff
@@ -0,0 +1,347 @@
...
+      .attr('height', svgHeight + margin.top + margin.bottom - 30)
+      .attr('width', svgWidth + margin.left + margin.right)
+      .append('g')
+      .attr('transform', 'translate(' + margin.left + ',' + margin.top + ')')
+      .selectAll('polygon')
+      .data(polygons3d)
+      .enter()
+      .append('polygon')
+        .attr('points', (d) => {
+            return d.points.map((d) => {
```
Comment (apryka) [inline lib/libraries/default/funnel-3d.js]: `d` is declared in upper scope

let's change it for `p` maybe so it should be

```.attr('points', (d) => {
            return d.points.map((p) => {
              return [p.x,p.y].join(',');
            })
            .join(' ');
        })```

File: docs/README.md [hunk 14]
```diff
@@ -42,6 +42,49 @@ const chart = new KeenDataviz({
...
+    //for funnel-3d:
+    effect3d: 'both-sides' // 'both-sides' - showing shades on both sides
+                           // 'left' - showing shades on left side
+                           // 'right' - showing shades on right side
+
+    //for horizontal-funnel-3d:
+    effect3d: 'both-sides' // 'both-sides' - showing shades on top and bottom
+                           // 'top' - showing shades on top
+                           // 'right' - showing shades on bottom
+  }
```
Comment (adamkasprowicz) [inline docs/README.md]: split those  effect3d  examples into 2 different code snippets

File: lib/index.js [hunk 15]
```diff
@@ -129,7 +129,19 @@ export const Dataviz = function (options = {}) {
...
    timezone: 'UTC',
    table: {},
-   renderOnVisible: false
+   renderOnVisible: false,
+   funnel: {
+     lines: true,
+     resultValues: true,
+     percents: {
+       show: false,
+       counting: 'relative',
```
Comment (adamkasprowicz) [inline lib/index.js]: absolute by default

File: lib/libraries/default/funnel-3d.js [hunk 16]
```diff
@@ -0,0 +1,347 @@
...
+          - margin.right - margin.left;
+    const svgHeight = funnelContainer.style('height').slice(0, -2)
+          - margin.top - margin.bottom - 30;
+    const elemHeight = svgHeight/(matrix.length - 1);
+    const yMarginElement = elemHeight * 0.3;
+    const shadowCut = (yMarginElement * 2.5);
+    let prevElemWidth = svgWidth;
+    let percent = (100).toFixed(decimals);
+
+    //normal funnel polygons preparing
```
Comment (adamkasprowicz) [inline lib/libraries/default/funnel-3d.js]: // prepare normal funnel polygons

**Pull request merged**